lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/driver/DefaultVortexMaster.java

```diff
@@ -140,4 +153,30 @@ public void taskletCancelled(final String workerId, final int taskletId) {
   public void terminate() {
     runningWorkers.terminate();
   }
+
+  private synchronized void putDelegate(final Tasklet tasklet, final VortexFutureDelegate delegate) {
+    taskletFutureMap.put(tasklet.getId(), delegate);
+  }
+
+  private synchronized VortexFutureDelegate fetchDelegate(final List<Integer> taskletIds) {
+    synchronized (taskletFutureMap) {
+      VortexFutureDelegate delegate = null;
+      for (final int taskletId : taskletIds) {
+        final VortexFutureDelegate currDelegate = taskletFutureMap.remove(taskletId);
+        if (currDelegate == null) {
+          // TODO[JIRA REEF-500]: Consider duplicate tasklets.
+          throw new RuntimeException("Tasklet should only be removed once.");
+        }
+
+        if (delegate == null) {
+          delegate = currDelegate;
+        } else {
+          assert delegate == currDelegate;
```

Comment:
Ideally, I would like the Future-matching logic to stay in the ``VortexMaster``, since all calls from the ``VortexWorker`` will need to be funneled through the ``VortexMaster``, and we would need to keep a map of taskletId to ``Future`` in either case. Perhaps I am misunderstanding something though, it would be great if you have found a better pattern to use here.